File: dbflow-rx2/src/main/java/com/raizlabs/android/dbflow/rx2/language/CursorResultFlowable.java
package com.raizlabs.android.dbflow.rx2.language;

import android.support.annotation.NonNull;
import android.support.annotation.VisibleForTesting;

import com.raizlabs.android.dbflow.config.FlowLog;
import com.raizlabs.android.dbflow.list.FlowCursorIterator;
import com.raizlabs.android.dbflow.sql.language.CursorResult;

import org.reactivestreams.Subscriber;
import org.reactivestreams.Subscription;

import java.util.concurrent.atomic.AtomicLong;

import io.reactivex.Flowable;
import io.reactivex.SingleObserver;
import io.reactivex.disposables.Disposable;

/**
 * Description: Wraps a {@link RXModelQueriable} into a {@link Flowable}
 * for each element represented by the query.
 */
public class CursorResultFlowable<T> extends Flowable<T> {


    @NonNull
    private final RXModelQueriable<T> modelQueriable;

    public CursorResultFlowable(@NonNull RXModelQueriable<T> modelQueriable) {
        this.modelQueriable = modelQueriable;

    }

    @Override
    protected void subscribeActual(final Subscriber<? super T> subscriber) {
        subscriber.onSubscribe(new Subscription() {
            @Override
            public void request(final long n) {
                modelQueriable.queryResults().subscribe(new CursorResultObserver<>(subscriber, n));
            }

            @Override
            public void cancel() {

            }
        });
    }

    @VisibleForTesting
    static class CursorResultObserver<T> implements SingleObserver<CursorResult<T>> {

        private final Subscriber<? super T> subscriber;
        private final long count;
        private final AtomicLong emitted;
        private final AtomicLong requested;
        private Disposable disposable;

        CursorResultObserver(Subscriber<? super T> subscriber, long count) {
            this.subscriber = subscriber;
            this.count = count;
            requested = new AtomicLong();
            emitted = new AtomicLong();
        }

        @Override
        public void onSubscribe(Disposable disposable) {
            this.disposable = disposable;
        }

        @Override
        public void onSuccess(CursorResult<T> ts) {
            long limit = this.count;
            int starting = limit == Long.MAX_VALUE && requested.compareAndSet(0, Long.MAX_VALUE)
                ? 0 : emitted.intValue();
            while (limit > 0) {
                FlowCursorIterator<T> iterator = ts.iterator(starting, limit);
                try {
                    long i = 0;
                    while (!disposable.isDisposed() && iterator.hasNext() && i++ < limit) {
                        subscriber.onNext(iterator.next());
                    }
                    emitted.addAndGet(i);
                    // no more items
                    if (!disposable.isDisposed() && i < limit) {
                        subscriber.onComplete();
                        break;
                    }
                    limit = requested.addAndGet(-limit);
                } catch (Exception e) {
                    FlowLog.logError(e);
                    subscriber.onError(e);
                } finally {
                    try {
                        iterator.close();
                    } catch (Exception e) {
                        FlowLog.logError(e);
                        subscriber.onError(e);
                    }
                }
            }
        }

        @Override
        public void onError(Throwable e) {
            subscriber.onError(e);
        }

    }
}

File: dbflow-rx2/src/main/kotlin/com/raizlabs/dbflow5/rx2/query/CursorResultFlowable.kt
package com.raizlabs.dbflow5.rx2.query

import com.raizlabs.dbflow5.config.DBFlowDatabase
import com.raizlabs.dbflow5.config.FlowLog
import com.raizlabs.dbflow5.query.CursorResult
import com.raizlabs.dbflow5.query.ModelQueriable
import com.raizlabs.dbflow5.rx2.transaction.asSingle
import com.raizlabs.dbflow5.transaction.Transaction
import io.reactivex.Flowable
import io.reactivex.SingleObserver
import io.reactivex.disposables.Disposable
import org.reactivestreams.Subscriber
import org.reactivestreams.Subscription
import java.util.concurrent.atomic.AtomicLong

/**
 * Description: Wraps a [ModelQueriable] into a [Flowable] that emits each item from the
 * result of the [ModelQueriable] one at a time.
 */
class CursorResultFlowable<T : Any>(private val modelQueriable: ModelQueriable<T>,
                                    private val database: DBFlowDatabase)
    : Flowable<T>() {

    override fun subscribeActual(subscriber: Subscriber<in T>) {
        subscriber.onSubscribe(object : Subscription {
            private var transaction: Transaction<CursorResult<T>>? = null

            override fun request(n: Long) {
                val single = database
                    .beginTransactionAsync { modelQueriable.queryResults(it) }
                    .asSingle()
                transaction = single.transaction
                single.subscribe(CursorResultObserver(subscriber, n))
            }

            override fun cancel() {
                transaction?.cancel()
            }
        })
    }

    internal class CursorResultObserver<T : Any>(
        private val subscriber: Subscriber<in T>, private val count: Long)
        : SingleObserver<CursorResult<T>> {
        private val emitted: AtomicLong = AtomicLong()
        private val requested: AtomicLong = AtomicLong()
        private var disposable: Disposable? = null

        override fun onSubscribe(disposable: Disposable) {
            this.disposable = disposable
        }

        override fun onSuccess(ts: CursorResult<T>) {
            val starting = when {
                this.count == Long.MAX_VALUE
                    && requested.compareAndSet(0, Long.MAX_VALUE) -> 0
                else -> emitted.toLong()
            }
            var limit = this.count + starting

            while (limit > 0) {
                val iterator = ts.iterator(starting, limit)
                try {
                    var i: Long = 0
                    while (!disposable!!.isDisposed && iterator.hasNext() && i++ < limit) {
                        subscriber.onNext(iterator.next())
                    }
                    emitted.addAndGet(i)
                    // no more items
                    if (disposable?.isDisposed == false && i < limit) {
                        subscriber.onComplete()
                        break
                    }
                    limit = requested.addAndGet(-limit)
                } catch (e: Exception) {
                    FlowLog.logError(e)
                    subscriber.onError(e)
                } finally {
                    try {
                        iterator.close()
                    } catch (e: Exception) {
                        FlowLog.logError(e)
                        subscriber.onError(e)
                    }
                }
            }
        }

        override fun onError(e: Throwable) {
            subscriber.onError(e)
        }

    }
}
